garak/payloads.py

```diff
@@ -0,0 +1,143 @@
+# SPDX-FileCopyrightText: Portions Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import logging
+import pathlib
+from typing import Generator, List, Union
+
+import garak._config
+import garak.exception
+
+
+class PayloadGroup:
+    """Represents a configured group of payloads for use with garak
+    probes. Each group should have a name, one or more payload types, and
+    a number of payload entries"""
+
+    def _load(self):
+        logging.debug("Loading payload %s from %s", self.name, self.path)
+        try:
+            with open(self.path, "r", encoding="utf-8") as payloadfile:
+                loaded_payload = json.load(payloadfile)
+
+        except FileNotFoundError as fnfe:
+            msg = "Payload file not found:" + str(self.path)
+            logging.error(msg, exc_info=fnfe)
+            raise garak.exception.PayloadFailure(msg) from fnfe
+
+        except json.JSONDecodeError as jde:
+            msg = "Payload JSON error:" + str(jde)
+            logging.error(msg, exc_info=jde)
+            raise garak.exception.PayloadFailure("Payload JSON error") from jde
+
+        self.types = loaded_payload["payload_types"]
+        self.payloads = [str(p) for p in loaded_payload["payloads"]]
+
+        self.detector_name = None
+        if "detector_name" in loaded_payload:
+            self.detector_name = str(loaded_payload["detector_name"])
+
+        self.detector_params = None
+        if "detector_params" in loaded_payload:
+            try:
+                self.detector_params = dict(loaded_payload["detector_params"])
+
+            except TypeError as te:
+                msg = "Payload detector_params must be a dict or empty, got: " + repr(
+                    loaded_payload["detector_params"]
+                )
+                logging.warning(msg, exc_info=te)
+                # raise ValueError(msg) from te
+
+        self._loaded = True
+
+    def __init__(self, name, path) -> None:
+        self.name = name
```

Comment:
@jmartin-tech is advocating for keying on path instead of name, which would provide a route to dropping name being used twice. But there'll still be a constructor var used to set an instance var. Where do you think a good place is to validate?

Reply:
If we're not using name twice, we're probably ok just to validate the path and let it live. Otherwise, we should probably check the validity of the name prior to assignment.

Reply:
After changing the payload key to be the filename stem, there's a lot of validation around loading that should get name right - and within this `PayloadGroup` class it's basically decorative. Have cast `name` to string in the constructor just for safety's sake.